drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2018 Philémon Jaermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <i2c.h>
+#include <init.h>
+#include <sensor.h>
+#include <logging/sys_log.h>
+
+#include "lsm303dlhc_magneto.h"
+
+static int lsm303dlhc_sample_fetch(struct device *dev,
+		enum sensor_channel chan)
+{
+	struct lsm303dlhc_magneto_data *drv_data = dev->driver_data;
+	u8_t tries = 11;
+	u8_t magneto_buf[6];
+	u8_t status;
+
+	/* Check data ready flag */
+	while (tries-- > 0) {
+		if (i2c_reg_read_byte(drv_data->i2c,
+					LSM303DLHC_MAGN_I2C_ADDR,
+					LSM303DLHC_SR_REG_M,
+					&status) < 0) {
+			SYS_LOG_ERR("Failed to read status register.");
+			return -EIO;
+		}
+
+		if ((status & LSM303DLHC_MAGN_DRDY) || tries == 0) {
+			break;
+		}
+
+		k_sleep(100);
+	}
+
+	if (!(status & LSM303DLHC_MAGN_DRDY)) {
+		SYS_LOG_ERR("Sensor data not available.");
+		return -EIO;
+	}
+
+	if (i2c_burst_read(drv_data->i2c,
+				LSM303DLHC_MAGN_I2C_ADDR,
+				LSM303DLHC_REG_MAGN_X_LSB,
+				magneto_buf, 6) < 0) {
+		SYS_LOG_ERR("Could not read magneto axis data.");
+		return -EIO;
+	}
+
+	drv_data->magneto_x = (magneto_buf[0] << 8) | magneto_buf[1];
+	drv_data->magneto_y = (magneto_buf[4] << 8) | magneto_buf[5];
+	drv_data->magneto_z = (magneto_buf[2] << 8) | magneto_buf[3];
+
```

Comment:
Is the above sequence correct?
Looking at the datasheet seems that magneto_buf[0] is the LSB
and magneto_buf[1] is the MSB. And so on. So you are populating
the samples in the wrong order. Does it make sense to you?

Reply:
I think you looked at the acceleration register. In the case of the acceleration, the LSB is at lower address, but for the magnetometer the MSB comes first (cf table 17)
Accel: OUT_X_L_A then OUT_X_M_A
Magn: OUT_X_H_M then OUT_X_L_M

Reply:
Correct, sorry! I have probably mis-looked at the datasheet.